### PR TITLE
Fix ambiguous reference when aliasing in combination with `ORDER BY`

### LINKED ISSUE #8391: SQL: Ambiguous reference when aliasing in combination with ORDER BY
### Describe the bug

When using an un-aliased name in an `ORDER BY` clause, while that name is aliased to the same name in a projection, datafusion raises an `AmbiguousReference` error.

If the name is not manually aliased, or the aliased name is used, no error is raised.

### To Reproduce

datafusion-cli:
```
❯ CREATE TABLE t (a INTEGER);
0 rows in set. Query took 0.003 seconds.
❯ SELECT t.a AS a FROM t ORDER BY t.a;
Schema error: Schema contains qualified field name t.a and unqualified field name a which would be ambiguous
```

The following variations do not raise an error:
```sql
SELECT a FROM t ORDER BY a;
SELECT a FROM t ORDER BY t.a;
SELECT t.a FROM t ORDER BY a;
SELECT t.a FROM t ORDER BY t.a;
SELECT t.a AS a FROM t ORDER BY a;
```

### Expected behavior

The result of the query `SELECT t.a AS a FROM t ORDER BY t.a` should be the same as from, for example `SELECT t.a AS a FROM t ORDER BY a` and `SELECT t.a FROM t ORDER BY t.a`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Asura7969**
## Which issue does this PR close?

Closes #8391.

## Rationale for this change

```sql
❯ CREATE TABLE t (a INTEGER);
0 rows in set. Query took 0.003 seconds.
❯ SELECT t.a AS a FROM t ORDER BY t.a;
Schema error: Schema contains qualified field name t.a and unqualified field name a which would be ambiguous
```

## What changes are included in this PR?

When the field alias is consistent with the original name, cancel the alias

## Are these changes tested?

Overwrite existing tests

### CODE REVIEW

File: datafusion/sql/src/select.rs [hunk 1]
```diff
@@ -384,7 +384,11 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     &[&[plan.schema()]],
                     &plan.using_columns()?,
                 )?;
-                let expr = col.alias(self.normalizer.normalize(alias));
+                let name = self.normalizer.normalize(alias);
+                let expr = match &col {
```
Comment (alamb) [inline datafusion/sql/src/select.rs:388]: Can you please add a comment here about what this is doing? I think it is avoiding adding an alias if the column name is the same. 

Do we have to worry about relation names here too (like what if `column.relation` is non null?)

File: datafusion/sql/tests/sql_integration.rs [hunk 2]
```diff
@@ -3546,9 +3546,9 @@ fn test_select_unsupported_syntax_errors(#[case] sql: &str, #[case] error: &str)
 fn select_order_by_with_cast() {
     let sql =
         "SELECT first_name AS first_name FROM (SELECT first_name AS first_name FROM person) ORDER BY CAST(first_name as INT)";
-    let expected = "Sort: CAST(first_name AS first_name AS Int32) ASC NULLS LAST\
-                        \n  Projection: first_name AS first_name\
-                        \n    Projection: person.first_name AS first_name\
+    let expected = "Sort: CAST(person.first_name AS Int32) ASC NULLS LAST\
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:3549]: this certainly looks better

**Review by alamb — COMMENTED** [review]
Thank you @Asura7969 . This is a really nice fix 🕵️ 💯  

The code in this PR looks good to me, though I do wonder abut the `relation` field (see inline comment)

Also, I tested the original query locally and this PR does indeed fix the problem 🎉 

```
DataFusion CLI v33.0.0
❯ CREATE TABLE t (a INTEGER);
0 rows in set. Query took 0.018 seconds.

❯ SELECT t.a AS a FROM t ORDER BY t.a;
0 rows in set. Query took 0.011 seconds.

❯
```

Can you please add some tests to ensure we don't break it accidentally in the future (the only tests that were changed are related to subqueries and I worry someone will accidentally break remove the fix code and not realize we introduced a regression)?

These should probably be end to end sqllogic tests:
https://github.com/apache/arrow-datafusion/tree/main/datafusion/sqllogictest

Perhaps added to https://github.com/apache/arrow-datafusion/blob/main/datafusion/sqllogictest/test_files/select.slt

**Comment by alamb** [comment]
fyi @Jesse-Bakker

File: datafusion/sql/src/select.rs (hunk 1 above)
Comment (Asura7969) [reply datafusion/sql/src/select.rs:388]: I understand, but I didn't expect the scenario where `column.relation` is inconsistent🤔

Comment (alamb) [reply datafusion/sql/src/select.rs:389]: I was thinking of a query like

```
EXPLAIN select a as a FROM table1 CROSS JOIN table2 order by a
```

Where both table1 and table2 have a column named `a`. I think we should add a test for this case too -- I'll try and do so later today

I would expect the test to fail

**Review by alamb — COMMENTED** [review]
Thanks @Asura7969 -- there is just one last thing I worry about with multiple relation names that I think we can resolve with some more tests.

File: datafusion/sql/src/select.rs (hunk 1 above)
Comment (Asura7969) [reply datafusion/sql/src/select.rs:389]: as you expected
```sql
DataFusion CLI v33.0.0
❯ CREATE TABLE table1 (a int) as values(1);
0 rows in set. Query took 0.037 seconds.

❯ CREATE TABLE table2 (a int) as values(1);
0 rows in set. Query took 0.012 seconds.

❯ EXPLAIN select a as a FROM table1 CROSS JOIN table2 order by a;
Schema error: Ambiguous reference to unqualified field a
```

**Review by alamb — APPROVED** [review]
Thank you for sticking with this @Asura7969 -- looks good to me

**Pull request merged**